src/coreComponents/constitutive/fluid/multifluid/constant/InvariantImmiscibleFluid.hpp

```diff
@@ -0,0 +1,259 @@
+                                                       PhaseProp::SliceType const phaseInternalEnergy,
+                                                       PhaseComp::SliceType const phaseCompFraction,
+                                                       FluidProp::SliceType const totalDensity ) const
+{
+  // Note: pressure and temperature are marked as unused as this model uses constant properties
+  GEOS_UNUSED_VAR( pressure, temperature );
+
+  using Deriv = constitutive::multifluid::DerivativeOffset;
+
+  integer nPhase = phaseDensity.value.size();
+  integer nComp = phaseCompFraction.value.size( 1 );
+
+  for( integer ip=0; ip<nPhase; ++ip )
+  {
+    // Phase fractions are equal to the composition for each phase
+    phaseFraction.value[ip] = composition[ip];
+    phaseFraction.derivs[ip][Deriv::dP] = 0.0;
+    for( integer ic = 0; ic < nComp; ++ic )
+    {
+      phaseFraction.derivs[ip][Deriv::dC+ic] = (ip == ic) ? 1.0 : 0.0;
+    }
+
+    // densities and viscosities constant
+    real64 const mult = m_useMass ? 1.0 : 1.0 / m_componentMolarWeight[ip];
+    phaseDensity.value[ip] = m_densities[ip] * mult;
+    phaseMassDensity.value[ip] = m_densities[ip];
+    phaseViscosity.value[ip] = m_viscosities[ip];
+    // derivatives
+    for( integer ic = 0; ic < nComp; ++ic )
+    {
+      phaseDensity.derivs[ip][Deriv::dC+ic] = 0.0;
+      phaseMassDensity.derivs[ip][Deriv::dC+ic] = 0.0;
+      phaseViscosity.derivs[ip][Deriv::dC+ic] = 0.0;
+    }
+
+    // zero enthalpy/internal energy
+    phaseEnthalpy.value[ip] = 0.0;
+    phaseInternalEnergy.value[ip] = 0.0;
+    for( integer d=0; d<phaseEnthalpy.derivs[ip].size( 0 ); ++d )
+    {
+      phaseEnthalpy.derivs[ip][d] = 0.0;
+      phaseInternalEnergy.derivs[ip][d] = 0.0;
+    }
+
+    // phase composition identity: each phase pure component
+    for( integer ic=0; ic<nComp; ++ic )
+    {
+      phaseCompFraction.value[ip][ic] = (ic==ip ? 1.0 : 0.0);
+      for( integer d=0; d<phaseCompFraction.derivs[ip][ic].size( 0 ); ++d )
+        phaseCompFraction.derivs[ip][ic][d] = 0.0;
+    }
+  }
+
+  // total density: sum over phases of phaseMassDensity * phaseFraction
+  computeTotalDensity( phaseFraction,
+                       phaseDensity,
+                       totalDensity );
+}
+
+GEOS_HOST_DEVICE
+GEOS_FORCE_INLINE
+void InvariantImmiscibleFluid::KernelWrapper::update( localIndex const k,
+                                                      localIndex const q,
+                                                      real64 const pressure,
+                                                      real64 const temperature,
+                                                      arraySlice1d< real64 const, compflow::USD_COMP - 1 > const & composition ) const
+{
+  // Create slice views of the member variables at cell k, quadrature point q
+  PhaseProp::SliceType phaseFractionSlice{m_phaseFraction.value[k][q], m_phaseFraction.derivs[k][q]};
```

Comment:
` m_phaseFraction( k, q )` should give the slice directly. No need to index the values and derivatives separately.